### PR TITLE
planner: support using hint to create hypo-indexes

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #43817

Problem Summary: planner: support using hint to create hypo-indexes

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: pkg/planner/optimize.go [hunk 1]
```diff
@@ -600,6 +605,27 @@ func setVarHintChecker(varName, hint string) (ok bool, warning error) {
 	return true, warning
 }
 
+func hypoIndexChecker(is infoschema.InfoSchema) func(db, tbl model.CIStr, cols ...model.CIStr) error {
+	return func(db, tbl model.CIStr, cols ...model.CIStr) error {
+		t, err := is.TableByName(db, tbl)
+		if err != nil {
+			return errors.NewNoStackErrorf("table '%v.%v' doesn't exist", db, tbl)
+		}
+		for _, col := range cols {
```
Comment (hawkingrei) [inline pkg/planner/optimize.go:614]: use ```slices.Contains``` to simple this code.

Comment (qw4990) [reply pkg/planner/optimize.go:616]: done

File: pkg/planner/optimize.go [hunk 2]
```diff
@@ -16,8 +16,10 @@ package planner
 
 import (
 	"context"
+	"github.com/pingcap/tidb/pkg/table"
```
Comment (AilinKid) [inline pkg/planner/optimize.go:19]: regroup 的 import

File: pkg/util/hint/hint.go [hunk 3]
```diff
@@ -312,6 +330,40 @@ func ParseStmtHints(hints []*ast.TableOptimizerHint,
...
+			}
+			tbl := hint.Tables[0].TableName
+			idx := hint.Tables[1].TableName
+			var colNames []model.CIStr
+			var cols []*model.IndexColumn
+			for i := 2; i < len(hint.Tables); i++ {
+				colNames = append(colNames, hint.Tables[i].TableName)
+				cols = append(cols, &model.IndexColumn{
+					Name:   hint.Tables[i].TableName,
+					Offset: i - 2,
```
Comment (AilinKid) [inline pkg/util/hint/hint.go:352]: should it align to the real column offset in tableInfo？

Comment (qw4990) [reply pkg/util/hint/hint.go:352]: No, the offset here means its offset in the index.

**Review by AilinKid — APPROVED** [review]

**Review by D3Hunter — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**